### PR TITLE
Basic functionality for the status API

### PR DESCRIPTION
**Pull request by stevex86**
# Go online
Go online
<hr>

# Check online friends
Friends who are online
<hr>

No JSON body required, just validate the token and use the user ID in the token to perform the work.

### CODE REVIEW

**Comment by nfegard** [comment]
I think /get_friends/ could be confusing but I am having trouble thinking of a better route. Maybe one of these: 

.../api/status/poll_friends/
.../api/status/sync/
.../api/status/get/friends

File: index.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+indexes:
```
Comment (nfegard) [inline index.yaml]: What does this file do? (Just curious)

File: service.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+// OnlineFriendModel serves as a model for each online friend returned by
+// this microservice.
+type OnlineFriendModel struct {
+	Username string `json:"username"`
+	UUID     string `json:"friendship_id"`
+	Online   bool   `json:"status"`
+}
+
```
Comment (nfegard) [inline service.go]: Just something to think about: Are we planning on making tags part of this microservice? Thinking about it now, it seems like a perfect fit but we may come up with reasons to separate it. Just food for thought.

File: service.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+func init() {
+	m := martini.Classic()
+	m.Use(middleware.AppEngineContext)
+	m.Use(middleware.AddHeaderJSON)
+	m.Post("/api/status/go_online/", middleware.Authenticate, GoOnline)
+	m.Get("/api/status/get_friends/", middleware.Authenticate, GetOnlineFriends)
```
Comment (nfegard) [inline service.go]: I personally hate underscores in routes. Are you particularly attached to them? (Just curious)

File: service.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+		KeysOnly()
+	for it := q.Run(ctx); ; {
+		key, err := it.Next(nil)
+		return key, err
+	}
+}
+
+func getOnlineFriends(ctx context.Context, userKey *datastore.Key) []OnlineFriendModel {
+	friendships := make([]schema.Friendship, 0)
+	q := datastore.NewQuery("Friendship").
```
Comment (nfegard) [inline service.go]: How do you feel about renaming q to query? I know it says NewQuery right after it lol.

File: service.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+	friends := make([]schema.User, len(friendKeys))
+	datastore.GetMulti(ctx, friendKeys, friends)
+
+	friendUsernames := make([]string, len(friendKeys))
+	for index, value := range friends {
+		friendUsernames[index] = value.Username
+	}
+
+	cachedItems, _ := memcache.GetMulti(ctx, friendUsernames)
```
Comment (nfegard) [inline service.go]: What does items stand for in cachedItems?

File: service.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	_, err := getUserKeyByName(ctx, t.UserIdentifier)
+	var response *juniper.Response
+	if err != nil {
+		e := &juniper.Error{
+			Code:        "temp",
+			Description: err.Error()}
+		response = &juniper.Response{
+			HTTPStatus: 500,
+			Err:        e,
+			Version:    "0.1.0"}
```
Comment (nfegard) [inline service.go]: Version: version

(at top)
const version string = "v 0.1.0"

We have been using "v 0.1.0" we can drop the "v" if doing so is more idiomatic. What do you think?

File: service.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+				Code:        "temp",
+				Description: err.Error()}
+			response = &juniper.Response{
+				HTTPStatus: 500,
+				Err:        e,
+				Version:    "0.1.0"}
+		} else {
+			response = &juniper.Response{
+				HTTPStatus: 200,
+				Err:        nil,
```
Comment (nfegard) [inline service.go]: do we need to set Err to nil here?

File: index.yaml (hunk 1 above)
Comment (stevex86) [inline index.yaml]: Really not sure why it even exists. It autogenerates database indexes that the application uses, but it doesn't need to be in source control.

File: service.go (hunk 2 above)
Comment (stevex86) [inline service.go]: I would imagine so

File: service.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+type OnlineFriendModel struct {
+	Username string `json:"username"`
+	UUID     string `json:"friendship_id"`
+	Online   bool   `json:"status"`
+}
+
+func init() {
+	m := martini.Classic()
+	m.Use(middleware.AppEngineContext)
+	m.Use(middleware.AddHeaderJSON)
```
Comment (stevex86) [inline service.go]: It only takes one Handler as an argument.

File: service.go (hunk 3 above)
Comment (stevex86) [inline service.go]: I also idomatically dislike underscores. I think I'm erring more towards `/api/status/friends/poll`

File: service.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+		models[index] = friendModel
+	}
+	return models
+}
+
+// GetOnlineFriends returns a repsonse containing a JSON array of information
+// about all friends of a user who are online
+func GetOnlineFriends(ctx context.Context, t *token.Token) (int, string) {
+	key, err := getUserKeyByName(ctx, t.UserIdentifier)
+	var response *juniper.Response
```
Comment (tforrest) [inline service.go]: They have been https://github.com/BrambleLLC/register/blob/develop/service.go#L111

File: service.go (hunk 5 above)
Comment (stevex86) [inline service.go]: Memcache is a collection of key to item pairs. Google has decided on the very descriptive name of memcache.Item as their terminology for items stored in cache.

File: service.go (hunk 9 above)
Comment (stevex86) [inline service.go]: Is there a reason for them not to be pointers? IMO it doesn't matter.

File: service.go (hunk 6 above)
Comment (stevex86) [inline service.go]: I request that we please drop the "v"

File: service.go (hunk 8 above)
Comment (nfegard) [inline service.go]: I see. I made changes like that to friends and Thomas did it in the other services too.

File: service.go (hunk 5 above)
Comment (nfegard) [inline service.go]: Haha gotcha.

**Comment by nfegard** [comment]
LGTM

**Pull request merged**